trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/DataFileDelimiter.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Drupal\trpcultivate_phenotypes\TripalCultivateValidator\ValidatorTraits;
+
+/**
+ * Provides setters focused for setting a delimiter used by the importer
+ * to delimit data/values in the data file, and getter to retrieve the set value.
+ */
+trait DataFileDelimiter {
```

Comment:
We just had a great in-person discussion regarding this trait that I am going to summarize here.

Action Items (on hold until after split row PR is merged):
1. This trait is going to disappear and the functionality is going to move into the fileTypes trait.
2. The fileTypes trait will have a mapping of MimeType => delimiter (e.g. 'text/tab-separated-values' => "\t") stored in a static variable.
3. There will be a getSupportedFileDelimiters() method that returns the list of delimiters supported by the set mime-types to use in validators.
4. The splitRowIntoColumns(string $line, string $mime_type) static method will use the mapping in action item 2 to split the line based on the mime_type passed in.
5. There won't be a setDataFileDelimiter() as the supported delimiters will also be set in the setFileTypes() with lookup being based on the mime types.
6. Instead there will be a static method to get the delimiter for a specific mime type (this can be used by both the static `splitRowIntoColumns()` and the validator checking the line format using `validateRawRow()`).

Discussion:
When discussing the fact that we would like to support multiple delimiters, it became apparent that this is also heavily integrated with the file extensions and mime types. In order to ensure that someone doesn't set a delimiter that is not correct for the mimetype/extension set with another setters, we've decided to let the single setter set all 3 (i.e. extensions, mime types, delimiters).

The discussion also passed into how we would use the multiple delimiters to split the row. While we want to use the `str_getcsv()` method to split the row, this only supports a single delimiter. We could loop through the supported delimiter until we find one that "Works" but we're worried that would be error prone and also a performance concern. The solution to this is to have the importer tell the splitRowIntoColumns() the mime type of the current file (which is can grab easily via the Drupal File object) and then this method can split specifically based on the type of delimiter supported for this mime type. Also when we validate the file row format using the validateRawRow() we will be able to use the same approach (i.e. get the delimiter for the specific mime type and then confirm the row uses it).

Reply:
Completed